Makefile

```diff
@@ -0,0 +1,242 @@
+#    and Bob instances represent two traders who can make and take
+#    offers with one another. The Mediator instance represents a Bisq
+#    contributor who can help resolve any technical problems or disputes
+#    that come up between the two traders.
+#
+# STEP 3: Configure the mediator Bisq node. In order to make and take
+# offers, Alice and Bob will need to have a mediator and a refund agent
+# registered on the network. Follow the instructions below to complete
+# that process:
+#
+#  a) Go to the Account screen in the Mediator instance and press CMD+N
+#  and a popup will appear. Click 'Unlock' and then click 'Register' to
+#  register the instance as a mediator.
+#
+#  b) While still in the Account screen, press CMD+D and follow the same
+#  steps as above to register the instance as a refund agent.
+#
+# When the steps above are complete, your localnet should be up and
+# ready to use. You can now test in isolation all Bisq features and use
+# cases.
+#
+
+
+# Set up everything necessary for deploying your localnet. This is the
+# default target.
+setup: build localnet
+
+clean: clean-build clean-localnet
+
+clean-build:
+	./gradlew clean
+
+clean-localnet:
+	rm -rf localnet
+
+# Build all Bisq binaries and generate the shell scripts used to run
+# them in the targets below
+build:
+	./gradlew build
+
+# Unpack and customize a Bitcoin regtest node and Alice and Bob Bisq
+# nodes that have been preconfigured with a blockchain containing the
+# BSQ genesis transaction
+localnet:
+	# Unpack the old dao-setup.zip and move things around for more concise
+	# and intuitive naming. This is a temporary measure until we clean these
+	# resources up more thoroughly.
+	unzip docs/dao-setup.zip
+	mv dao-setup localnet
+	mv localnet/Bitcoin-regtest localnet/bitcoind
+	mv localnet/bisq-BTC_REGTEST_Alice_dao localnet/alice
+	mv localnet/bisq-BTC_REGTEST_Bob_dao localnet/bob
+	# Remove the preconfigured bitcoin.conf in favor of explicitly
+	# parameterizing the invocation of bitcoind in the target below
+	rm -v localnet/bitcoind/bitcoin.conf
+	# Avoid spurious 'runCommand' errors in the bitcoind log when nc
+	# fails to bind to one of the listed block notification ports
+	echo exit 0 >> localnet/bitcoind/blocknotify
+
+# Deploy a complete localnet by running all required Bitcoin and Bisq
+# nodes, each in their own named screen window. If you are not a screen
+# user, you'll need to run each of the make commands manually in a
+# separate terminal or as a background job.
+#
+# NOTE: You MUST already be attached to a screen session for the
+# following commands to work properly.
+deploy: setup
+	screen -t bitcoin make bitcoind
+	sleep 2    # wait for bitcoind rpc server to start
+	make block # generate a block to ensure Bisq nodes get dao-synced
+	screen -t seednode make seednode
+	screen -t seednode2 make seednode2
+	screen -t alice make alice
+	screen -t bob make bob
+	screen -t mediator make mediator
```

Comment:
This is close to what I have locally just in a .rc file. Does bisq not handle bitcoind not having the rpc server up first? I've never run into a problem, but I always start bitcoind first so maybe I just got lucky?

Reply:
Yes bitcoind must run first. If the seed runs as full DAO node bitcoind must run before the seed node.

Reply:
Yeah that is why in the original [bisq-workspace](https://github.com/bodymindarts/bisq-workspace) I start bitcoind with a script that doesn't complete [until bitcoind is responsive](https://github.com/bodymindarts/bisq-workspace/blob/master/scripts/start_bitcoind.sh#L39-L49).

Reply:
Aside from @chimp1984's point that bitcoind should run before the seednode in any case, the actual reason I introduced the `sleep 2` and `make block` entries here is because if a new block (block 112) doesn't get created prior to starting the desktop nodes, they spin forever on DAO synchronization (which is annoying to look at but also eats up CPU, spins up fans and makes everything feel heavyweight). I believe this is because the desktop nodes need to actually receive a block notification, but it might also be that they need to have at least one confirmation on the genesis tx in block 111, maybe both. In any case, without block 112, there is a strange status message in `DAO->BSQ wallet->transactions` that reads "Awaiting synchronization... Validated 111 of 0 blocks". So something is off there. To see this behavior for yourself, just comment out the `sleep` and `make block` lines and run through the makefile instructions as usual.

Reply:
That makes a lot of sense. I think it should however be the responsibility of the `bitcoind` command to spin it up in a way that is consistent for the other commands to consume.

When adding an alternative `deploy-tmux` command for example that logic has to be duplicated.

Is it still a problem if `block 112` isn't present when the desktop client sping up but appears very soon thereafter?

Reply:
> Is it still a problem if `block 112` isn't present when the desktop client is spun up but appears very soon thereafter?

It is a problem in the sense that if 112 doesn't show up quickly, the contributor using the makefile is going to hear their fans spin up as the dao synchronization wait loop eats up CPU (we should obviously try to solve that problem at the root, but in the meantime...).

So if we're talking about the difference between generating block 112 just after deploying bitcoind vs. doing it just after deploying all the nodes, i.e. a difference measured in (sub)seconds, then it's not a problem in practice. Doing something like @KanoczTomas's start_bitcoind wrapper (which I believe you used elsewhere, @bodymindarts) might be a solution, but I've avoided any such shell scripts and/or variables thus far with the intention of making everything that's being done absolutely clear to the reader because it's all in one place (the makefile), free of any indirections or abstractions they need to dig into.

Reply:
Yes I understand the intention with having it all self-contained in the makefile and like the approach.
It seems to be a trade trade-off between ease-of-use (via a wrapper script) and 'time searching for the exact commands being executed'. One is more focused on the developer who wants a streamlined workflow, the other on a first-timer just seeing how things work.
In general I like being explicit (ie. no wrapper or indirection) and try to achieve that whenever possible, but in my experience (and I use a Makefile in every single project I do) when there are more than a few commands being executed or there are other complications having 1 level of indirection, like:
```
something:
    scripts/do_something.sh
```
is an acceptable trade off, especially if it improves the dev-workflow (and active devs are probably the primary user of this interface). For first timers having to jump into the scripts folder for more information should be fine.

Another benefit is having the individual make commands closer together in the Makefile, which means not having to scroll so far when trying to assess what all the main commands are that are used to interface with the project as a dev. This also benefits first-time contributors.

Reply:
+1, thanks. Where stuff gets unwieldy and actually hurts comprehension, we should shell out to something under `scripts/`. The current `screen` voodoo is probably a good example of this. I'd like to try to keep the starting of bitcoind and Bisq nodes script-free, as I think there's value in seeing them all together, explicitly parameterized in one place as we have it now. Let's just do whatever makes sense over time though.